File: examples/federated-work/python_buyer/subcontract.py
"""Independent checks for one permitted specialist and its exact disclosure."""
import protocol as p

PROFILE = "chio.example.security-review-agreement.v4"


def validate_policy(parent):
    policy = p.fields(parent["subcontract"], ("specialist", "delegate", "paths", "inputSha256", "priceCeiling"))
    p.hex_bytes(policy["specialist"])
    p.hex_bytes(policy["delegate"])
    p.hex_bytes(policy["inputSha256"])
    p.require(len({parent["buyer"], parent["provider"], policy["specialist"], policy["delegate"]}) == 4, "subcontract authority and agent keys must be distinct")
    paths = policy["paths"]
    p.require(type(paths) is list and 1 <= len(paths) <= 16 and
              all(type(path) is str and path.startswith("/") and len(path.encode()) <= 256 for path in paths),
              "unsupported disclosure paths")
    p.require(paths == sorted(set(paths)), "disclosure paths must be sorted and unique")
    p.integer(policy["priceCeiling"], 100, 100)
    p.require(policy["priceCeiling"] <= parent["priceCeiling"], "subcontract exceeds parent ceiling")
    return policy


def project(source, paths):
    import review
    review.observations(source)
    document = p.load_json(source)
    selected, references = {}, set()
    for path in paths:
        p.require(path in document["paths"], "approved disclosure path is absent")
        operations = {}
        for method in review.METHODS & document["paths"][path].keys():
            operation = document["paths"][path][method]
            security = operation.get("security", document.get("security", []))
            for alternative in security:
                references.update(alternative)
            operations[method] = {"security": security}
        p.require(operations, "approved path has no supported operations")
        selected[path] = operations
    schemes = {}
    for name in references:
        original = document["components"]["securitySchemes"][name]
        fields = ("type", "scheme") if original["type"] == "http" else ("type", "name", "in")
        schemes[name] = {key: original[key] for key in fields}
    result = p.canonical({"openapi": "3.1.0", "paths": selected, "components": {"securitySchemes": schemes}}).decode()
    review.observations(result)
    return result


def disclosure(work):
    parent = work["acceptance"]["quote"]["agreement"]
    policy = validate_policy(parent)
    result = project(work["input"], policy["paths"])
    import hashlib
    p.require(hashlib.sha256(result.encode()).hexdigest() == policy["inputSha256"], "subcontract disclosure differs from approved bytes")
    return result


def child_agreement(parent):
    policy = validate_policy(parent)
    p.require(parent["profile"] == PROFILE and parent["subcontracting"] is True, "parent does not permit a subcontract")
    return {"profile": p.PROFILE, "jobId": "child-" + p.digest(parent), "inputSha256": policy["inputSha256"],
            "buyer": policy["delegate"], "provider": policy["specialist"], "priceCeiling": policy["priceCeiling"],
            "deadline": parent["deadline"], "checker": parent["checker"], "subcontracting": False, "creditProfile": p.CREDIT}


def verify_child(work, value):
    import review
    p.fields(value, ("request", "delivery"))
    expected = child_agreement(work["acceptance"]["quote"]["agreement"])
    p.require(p.same(value["request"]["acceptance"]["quote"]["agreement"], expected)
              and value["request"]["input"] == disclosure(work), "specialist evidence changes approved work or disclosure")
    quote = value["request"]["acceptance"]["quote"]
    p.require("subcontractPermit" in quote, "specialist evidence omitted procurement permit")
    permit = verify_permit(quote["subcontractPermit"], expected, work["acceptance"]["quote"]["agreement"]["provider"])
    p.require(p.same(permit, permit_terms(work)), "specialist permit changes the parent work request")
    rejected, _ = review.terminal(value["request"], value["delivery"], {name: expected[name] for name in ("buyer", "provider")})
    p.require(not rejected, "specialist did not deliver an accepted result")


def permit_terms(parent):
    child = child_agreement(parent["acceptance"]["quote"]["agreement"])
    return {"schema": "chio.example.subcontract-permit.v1",
            "parentAgreementSha256": p.digest(parent["acceptance"]["quote"]["agreement"]), "parentRequestSha256": p.digest(parent),
            "childAgreementSha256": p.digest(child), "delegate": child["buyer"], "specialist": child["provider"],
            "priceCeiling": child["priceCeiling"], "currency": "TST", "expiresAt": child["deadline"]}


def verify_permit(value, child, promisor):
    body = p.envelope(value, promisor)
    p.fields(body, ("schema", "parentAgreementSha256", "parentRequestSha256", "childAgreementSha256", "delegate",
                    "specialist", "priceCeiling", "currency", "expiresAt"))
    for name in ("parentAgreementSha256", "parentRequestSha256", "childAgreementSha256"):
        p.hex_bytes(body[name])
    p.require(body["schema"] == "chio.example.subcontract-permit.v1" and body["childAgreementSha256"] == p.digest(child)
              and child["profile"] == p.PROFILE and child["subcontracting"] is False and "subcontract" not in child
              and child["jobId"] == "child-" + body["parentAgreementSha256"]
              and child["buyer"] == body["delegate"] and child["provider"] == body["specialist"]
              and len({body["delegate"], body["specialist"], promisor}) == 3
              and body["priceCeiling"] == child["priceCeiling"] == 100 and body["currency"] == "TST"
              and body["expiresAt"] == child["deadline"], "permit does not authorize this child procurement")
    p.integer(body["priceCeiling"], 100, 100)
    p.integer(body["expiresAt"], 1)
    return body

File: examples/federated-work/src/subcontract.rs
pub mod enrollment;
pub mod permit;
#[cfg(test)]
mod tests;
pub mod worker;

// An exact disclosure and child-work contract selected by the original buyer.
use crate::{common::*, review};
use chio_core_types::PublicKey;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::{BTreeMap, BTreeSet};

pub const PROFILE: &str = "chio.example.security-review-agreement.v4";

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Policy {
    pub specialist: PublicKey,
    pub delegate: PublicKey,
    pub paths: Vec<String>,
    pub input_sha256: String,
    pub price_ceiling: u64,
}

impl Policy {
    pub fn validate(&self, parent: &Agreement) -> Result<()> {
        if self.specialist == parent.buyer
            || self.specialist == parent.provider
            || self.delegate == parent.buyer
            || self.delegate == parent.provider
            || self.delegate == self.specialist
            || self.paths.is_empty()
            || self.paths.len() > 16
            || self
                .paths
                .iter()
                .any(|path| !path.starts_with('/') || path.len() > 256)
            || self.paths.windows(2).any(|pair| pair[0] >= pair[1])
            || self.price_ceiling != 100
            || self.price_ceiling > parent.price_ceiling
            || self.input_sha256.len() != 64
            || !self
                .input_sha256
                .bytes()
                .all(|b| b.is_ascii_digit() || (b'a'..=b'f').contains(&b))
        {
            return Err(
                "subcontract policy exceeds the supported disclosure or spending boundary".into(),
            );
        }
        Ok(())
    }
}

/// Reconstruct only selected paths' effective authentication declarations.
/// Descriptions, parameters, examples, extensions and unselected paths never
/// enter the child document, even when nested beneath an approved path.
pub fn project(input: &str, paths: &[String]) -> Result<String> {
    review::check_openapi(input)?;
    let document: Value = serde_json::from_str(input)?;
    let mut selected = BTreeMap::new();
    let mut referenced = BTreeSet::new();
    for path in paths {
        let item = document["paths"]
            .get(path)
            .and_then(Value::as_object)
            .ok_or("approved path is absent")?;
        let mut operations = BTreeMap::new();
        for method in [
            "get", "put", "post", "delete", "options", "head", "patch", "trace",
        ] {
            let Some(operation) = item.get(method) else {
                continue;
            };
            let security = operation
                .get("security")
                .or_else(|| document.get("security"))
                .cloned()
                .unwrap_or_else(|| json!([]));
            for requirement in security.as_array().ok_or("security is not an array")? {
                for name in requirement
                    .as_object()
                    .ok_or("security requirement is not an object")?
                    .keys()
                {
                    referenced.insert(name.clone());
                }
            }
            operations.insert(method.to_string(), json!({"security":security}));
        }
        if operations.is_empty() {
            return Err("approved path has no supported operations".into());
        }
        selected.insert(path.clone(), operations);
    }
    let mut schemes = BTreeMap::new();
    for name in referenced {
        let original = &document["components"]["securitySchemes"][&name];
        let keys = if original["type"] == "http" {
            vec!["type", "scheme"]
        } else {
            vec!["type", "name", "in"]
        };
        let scheme = keys
            .into_iter()
            .map(|key| (key.to_string(), original[key].clone()))
            .collect::<BTreeMap<_, _>>();
        schemes.insert(name, scheme);
    }
    let projected =
        json!({"openapi":"3.1.0","paths":selected,"components":{"securitySchemes":schemes}});
    let bytes = chio_core_types::canonical_json_bytes(&projected)?;
    let text = String::from_utf8(bytes)?;
    review::check_openapi(&text)?;
    Ok(text)
}

pub fn child_agreement(parent: &Agreement) -> Result<Agreement> {
    let policy = parent
        .subcontract
        .as_ref()
        .ok_or("no subcontract permission")?;
    policy.validate(parent)?;
    if parent.profile != PROFILE || !parent.subcontracting {
        return Err("parent does not permit this subcontract".into());
    }
    Ok(Agreement {
        profile: WORK_PROFILE.into(),
        job_id: format!("child-{}", digest(parent)?),
        input_sha256: policy.input_sha256.clone(),
        buyer: policy.delegate.clone(),
        provider: policy.specialist.clone(),
        price_ceiling: policy.price_ceiling,
        deadline: parent.deadline,
        checker: parent.checker.clone(),
        subcontracting: false,
        subcontract: None,
        credit_profile: parent.credit_profile.clone(),
    })
}

pub fn disclosure(request: &review::ReviewRequest) -> Result<String> {
    let parent = &request.acceptance.quote.agreement;
    let policy = parent
        .subcontract
        .as_ref()
        .ok_or("no subcontract permission")?;
    policy.validate(parent)?;
    let input = project(&request.input, &policy.paths)?;
    if chio_core_types::sha256_hex(input.as_bytes()) != policy.input_sha256 {
        return Err("subcontract disclosure differs from buyer-approved bytes".into());
    }
    Ok(input)
}

#[derive(Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ChildDelivery {
    pub request: review::ReviewRequest,
    pub delivery: Value,
}

pub fn verify_child(parent: &review::ReviewRequest, value: &Value) -> Result<()> {
    let child: ChildDelivery = serde_json::from_value(value.clone())?;
    let expected = child_agreement(&parent.acceptance.quote.agreement)?;
    if digest(&child.request.acceptance.quote.agreement)? != digest(&expected)?
        || child.request.input != disclosure(parent)?
    {
        return Err(
            "specialist evidence changes the permitted child agreement or disclosure".into(),
        );
    }
    permit::verify_parent(parent, &child.request.acceptance.quote)?;
    child.request.validate(&Peers {
        buyer: expected.buyer,
        provider: expected.provider,
    })?;
    if review::verify_terminal(&child.request, &child.delivery)?.0 {
        return Err("specialist did not deliver an accepted result".into());
    }
    Ok(())
}
